### ISSUE TITLE
Toolbar buttons for all the widgets should be provided

### ISSUE THREAD

**Issue by danielo515** [Jermolene/TiddlyWiki5#2873]
Hello,

I really think that each widget should have an equivalent button on the toolbar. As a regular user of tiddlywiki I find myself going to tiddlywiki.com every time I have to create a button, a list widget, an edit field... there are common operations on tiddlywiki, but not common enough to remember the syntax by heart. Having buttons for each core widget makes a lot of sense, and will make life significantly easier to newcomers (and not so new!).

I am also ok if we decide that this should be better provided as snippets, but they should be provided one way or another.

Once a decision is taken, I am happy to make a pull request.

Regards

**Comment by pmario** [comment]
We have the "stamp" button, that can hold the templates. Those templates can be loaded as a plugins. It may be core plugins, with the ability to remove and replace them.

**Comment by danielo515** [comment]
@pmario when I said snippets I mean the stamp button.
So, you vote to include them as snippets right?
Of course I agree it should be a plugin, but as you said I think it should be a core plugin, included by default.

Regards

**Comment by Jermolene** [comment]
Thanks @danielo515 @pmario can we define what is meant by the category "core plugin"? it is not a term defined on tw.com, but is easily confused with **the** core plugin and so I'm not sure we should adopt the term. I think you mean a plugin that is included in empty.html? At the moment, empty.html only includes the core plugin itself (see the confusion?).

**Comment by pmario** [comment]
@Jermolene .. you are right. We actually should say `tiddlywiki plugin`, because it may be part of the `/plugins/tiddlywiki/` directory in the repo.

**Comment by pmario** [comment]
>I think you mean a plugin that is included in empty.html?

hmmm, If it includes a plugin, it's not empty anymore. ... May be empty.html should get an improved GettingStarted. So there may be deeplinks to the ControlPanel - Plugins tab. So new users can find them more easily. .... 

Also a link to the existing editions http://tiddlywiki.com/#Editions and http://tiddlywiki.com/#Community tiddlers may be an option, to improve discoverability. 

just some thoughts.

**Comment by Jermolene** [comment]
Good line of thought @pmario but I worry that ""tiddlywiki plugin"" might be confused for the category of ""all"" tiddlywiki plugins... Saying that a file is a "tiddlywiki plugin" would be particularly ambiguous.

There are alternative ways of describing things that do conform to established terms: "included the empty edition", for example is unambiguous and descriptive, albeit doesn't trip off the tongue. The acronym "ITEE" might work? Or "TWEEP" for "TiddlyWiki empty edition plugin". Or simplified to "this thing should be in the empty edition".

Relatedly, I think it would be desirable for tw.com to have a "big green download button" in the sidebar, with a dropdown that makes it explicit that the default is the "empty edition", and provides options for the other editions present on tw.com.

If we did come up with a term, we could use it as a tag for the tiddlers sent over by the plugin library, and then the "Get Plugins" modal could include a badge for plugins that are included in the empty edition, and a button to filter by that tag.

**Comment by pmario** [comment]
>If we did come up with a term, we could use it as a tag for the tiddlers sent over by the plugin library, and then the "Get Plugins" modal could include a badge for plugins that are included in the empty edition, and a button to filter by that tag.

We are a bit OT now, I think, we should create a new issue, to track this idea. I'm very much in favor of having some plugin-categories eg: "Import tools", "Developer Tools", "Edit helpers" (imo OP belongs here), ... We have 33 plugins at the moment, and only a view of them are useful for newbees. So it's hard to know, what to try out.

**Comment by Jermolene** [comment]
btw @pmario 

> If it includes a plugin, it's not empty anymore

On the other hand, it is empty of tiddlers from the users perspective; the "More/All" tab is empty. And even today there are several system tiddlers included in the empty edition; it has "never" been truly empty.

> we should create a new issue, to track this idea

Please do.

**Comment by danielo515** [comment]
>>If it includes a plugin, it's not empty anymore

>On the other hand, it is empty of tiddlers from the users perspective; the "More/All" tab is empty. And even today there are several system tiddlers included in the empty edition; it has "never" been truly empty.

I agree with @Jermolene here. I always though about the empty edition as a TW with no tiddlers. However what @pmario says makes sense. I may want a truly empty tiddlywiki edition, without any plugin except the core. But I also think that this is not desirable for beginners, where a plugin like the one were are discussing here will help them a lot. This makes me think that maybe is a good idea to keep the empty edition fully empty and create a new beginner edition with some helping plugins and make this edition the default one.

Regarding the emptiness of the empty edition, it also includes two themes, that are nothing but plugins.

Regards

**Comment by danielo515** [comment]
> We are a bit OT now
> imo OP belongs here

@pmario  I don't understand a word :smile:

However having the plugins organized in categories makes a lot of sense to me too.

**Comment by pmario** [comment]
OT .. off topic
OP .. original post :)

**Comment by twMat** [comment]
Funny discussion.

As for the term "empty TW", I think we should use the term "**standard TW**" instead.

...then refer to it as:

The "standard **distro**". This is an established term albeit tech lingo.

Alternatively; we already use the term "**edition**", thus "standard edition". That said, "edition" is *not* a good term because it means [this](https://en.wikipedia.org/wiki/Edition) for most people, i.e the *key* characteristic is that different editions are very *similar*!

IMO the best term would be [application](http://www.webopedia.com/TERM/A/application.html) (or [here](https://en.wikipedia.org/wiki/Application_software)), thus "the standard TW **application**"... or, as things are probably going anyway; the "standard TW app".

**As for the OP**, i.e an edit tool for stamping widgets, I agree it would be very useful.

**Comment by danielo515** [comment]
I think what @twMat said was in the same line of what I was trying to say when I talked about a beginner's edition. So I subscribe every word. The most easy is to start with tiddlywiki the better, and not having to go to tw.com every time you want to use a widget it's definitely an improvement, and will increase discoverability on widgets in an easier way than reading all the documentation of all the widgets.

Returning back to the topic, I want to know the opinion of other people about how this should be included. @pmario suggested to include it as part of the `stamp` collection, and I agree. Others may prefer to provide certain widgets (like the button) in their own toolbar button.

Once we decide how I'll open a pull request, then anyone should be able to provide their own recipes.

Regards

**Comment by twMat** [comment]
Here's a drawing

![image](https://user-images.githubusercontent.com/4534056/27131267-3b215104-510a-11e7-8a27-44472e4de0da.png)

The idea is a "latest" showing ~10 of the last widgets used because the full list is very long. It could show a prepopulated list to begin with.

Maybe vertical tabs are better?

**Comment by danielo515** [comment]
Vertical tabs takes horizontal space, which is a treasure on small screens, better keep it too tabs. @twMat could you highlight the button pressed ?

**Comment by twMat** [comment]
>Vertical tabs takes horizontal space, which is a treasure on small screens, better keep it too tabs.

Sure, no problem. I just figured that with clean names for the widgets, like "Button" rather than "ButtonWidget" then not much horizontal space is used. 

>could you highlight the button pressed ?

Not sure I understand? Do you want me to remake the illustration? (Note, I did not implement it!) The idea is simply a tool button with e.g a "$" sign, as illustrated above. (I actually pasted over the "more" tool there but that should of course not be done.) And I guess the clicked button should turn blue when clicked but I figured it was difficult to see the "$" sign then.

**Comment by danielo515** [comment]
Hello @twMat 

> The idea is simply a tool button with e.g a "$" sign, as illustrated above.

I suspect that the button with the dollar sign was the one you were proposing, however I was not sure if that was the button pressed or just a custom button of your own wiki that got captured on your screenshot.

> Do you want me to remake the illustration? 

That is what I meant, yes

**Comment by twMat** [comment]
@danielo515 - ok I've updated the image, and therefore also edited my post where I asked what you meant, so to not confuse readers. If you update your replies accordingly, I'll delete this post.

**Comment by twMat** [comment]
Come to think of it, I think the "$" is not a very good icon. It is scary and, more critically, it doesn't really have any meaning. And there is a stated ambition to hide it from the front UI.

Something indicating a "primitive building block" or similar would be more appropriate IMO. Perhaps something along LEGO blocks ([eg](https://cdn4.iconfinder.com/data/icons/unigrid-baby/60/012_blocks_bricks_lego_constructor_toy_baby-512.png) or [eg](https://cdn0.iconfinder.com/data/icons/computers-and-development/432/plugin-2-128.png)) or atoms [eg](http://www.iconninja.com/files/678/1021/94/atom-atomic-nuclear-icon.png).

**Comment by tobibeer** [comment]
> more critically, it doesn't really have any meaning

For those users using widgets it has all the right meaning. For those who don't, at some point, it's worth learning what widgets are and that this exists to facilitate adopting them.
